### DISCUSSION TITLE
Support multiple vercel.json files for monorepo architectures

### DISCUSSION THREAD

**Discussion by th-m** [vercel/vercel#7649] [Help] [answered]
I have a monorepo that has a React SPA  and a Next app. I would like to configure rewrites for my SPA that do not apply to my Next app. Since the vercel.json rewrites doesn't support domain names or env vars I don't have any way to do this.

**Comment by Joroze** [comment]
I'm surprised this isn't possible. @TurboRepo team? :)

**Comment by amyegan** [comment, marked as answer]
Thanks for reaching out! You should be able to put a `vercel.json` config in the root directory of each app. Is that not working for you? If not, can you share more about what errors you are seeing?

**Reply by Joroze** [reply]
Actually, thanks - I believe you're right. It does work. On the other hand, I believe the Vercel CLI outputs a message saying it's not supported.

**Reply by amyegan** [reply]
I'll have to share that with the CLI team. Thanks for letting us know!

**Comment by freddneos** [comment]
Any new about it ? I'm using nx.dev monorepo and facing the problems with it...

**Reply by amyegan** [reply]
Hi @freddneos! You can have a `vercel.json` config in each project, rather than just one at the root of the monorepo. If that is not working for you, can you share more error details so I can get a better understanding of what is going wrong?

**Reply by futantan** [reply]
hi @amyegan , for a monorepo like
```
my-monorepo
├─ apps
│  ├─ web
│  └─ doc
├─ packages
│  ├─ tsconfig
│  └─ shared-utils
└─ sdk
```
Do you mean add a `vercel.json` file under `apps/web` and `apps/doc`?
When I run `vercel`, can I run it at the root of the monorepo project?

**Reply by amyegan** [reply]
Yes, that's what I meant. Vercel CLI should always be run at the root of your [monorepo](https://vercel.com/docs/concepts/monorepos), and individual project `vercel.json` files will be picked up during the build.

**Reply by futantan** [reply]
thanks @amyegan, but it seems not working when `verson.json` is under `apps/web`, but it's working when it's at the root of the monorepo.
I'm only using `vercel` CLI to deploy, is there anything that I need to configure in vercel dashboard?

**Reply by amyegan** [reply]
Recommended configuration settings vary depending on the monorepo tools you are using, but you can use the [Turborepo](https://vercel.com/docs/concepts/monorepos/turborepo#), [Nx](https://vercel.com/docs/concepts/monorepos/nx#), or [Yarn Workspaces](https://vercel.com/guides/deploying-yarn-monorepos-to-vercel) guide to help with setup.

**Reply by danieljvdm** [reply]
In @futantan's example folder structure, if `apps/doc` isn't an app that I want to deploy on Vercel, how do I prevent Vercel trying to validate that the git author is added to the Vercel team? I have a large team and only a couple of the apps we're deploying to Vercel but every single commit checks Vercel/git authors and it fails every build. I don't want to add everyone to Vercel when they're not using it.

**Reply by amyegan** [reply]
Hi @danieljvdm! This sounds like a slightly different question. Would you might starting a [new discussion](https://github.com/vercel/community/discussions/new) with a description of your problem to make the conversation easier to read?

**Comment by fcastillo90** [comment]
@amyegan hi! im having a hard time making Vercel+Nx work with individual vercel.json files on each of the monorepo app. Could you please give assistance? my deploy triggers automatically by vercel git integration watching the repo. The config for one of the app is the following:

root directory: `.`
build command: `npx nx build landing`
output directory: `dist/apps/landing/.next`
install command: `yarn install`
ignore command: `npx nx-ignore landing`

but the vercel.json located at `apps/landing` is being ignored. This vercel.json file contains `env` vars, `headers` and `redirects` for that individual app

is there something I'm missing out?

thanks!

**Reply by fcastillo90** [reply]
it seems that for Nx projects this configuration is what does the trick for vercel to pick the individual vercel.json of the app:

root directory: `apps/landing`
build Command: `cd ../../ && npx nx build landing`
ignore Command: `cd ../../ && npx nx-ignore landing`
install Command: `yarn install`
output Directory: `../../dist/apps/landing/.next`

is this the recommended config?

Thanks!

**Reply by iceStorm** [reply]
Many thanks for your solution. This helped me deploy a NestJS application along with a ReactJS app composed by Nx Monorepo.

**Comment by MarcStdt** [comment]
> it seems that for Nx projects this configuration is what does the trick for vercel to pick the individual vercel.json of the app:
> 
> root directory: `apps/landing` build Command: `cd ../../ && npx nx build landing` ignore Command: `cd ../../ && npx nx-ignore landing` install Command: `yarn install` output Directory: `../../dist/apps/landing/.next`
> 
> is this the recommended config?
> 
> Thanks!

Thanks this helped a lot, but I still think it's an imperfect solution.
I am deploying a SPA using a nx monorepo, so I need to use the rewrite function of the `vercel.json`.
It would be nice to have a better solution than manually switching directory.
